### PR TITLE
30-Layer Recursive GPT (10 Physical x 3 Recursive) + VRL + Parallel Muon

### PR DESCRIPTION
**Pull request by Jtss-ux**
Extreme 30-layer virtual depth via 3x recursion (10 physical layers) + Value Residual Learning (VRL) + Multi-Token Prediction (MTP) + BigramHash(3072) + Trigram + Parallel Muon. Optimized for H100 with Activation Checkpointing. Target Score: 1.10 BPB. Compressed Size: <16MB target.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — 30-Layer Recursive GPT (10 Physical x 3 Recursive) + VRL + Parallel Muon

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## PR #1269 Audit — head SHA 494c8aec8cbb89213989a7367ebaa6fdcbc8c145 ### Summary PR adds three submission record folders plus audit/tooling scripts from author Jtss-ux. The primary train_gpt.py (root level) is a ternary GPT with Muon optimizer. No illegal patterns detected. ### Check 1: ILLEGAL n-gram family bug (target XOR'd into hash key) CLEAN. The bigram hash is implemented at lines 869–871: ```python prev = F.pad(input_ids[:, :-1], (1, 0), value=0) x = x + self.bigram_emb(prev).float() ``` Key used for embedding lookup is `prev` — the *previous* token (input context), not the target. Target tokens (`target_ids`) are never used as a hash key or embedding index during forward. This is a standard causal bigram context embedding, not an illegal n-gram family leakage. ### Check 2: ILLEGAL Pre-Quant TTT (multi-epoch on val_tokens without score-first) CLEAN. No TTT of any kind present. The training loop (lines 1245–1358) runs on `train_loader` only. `val_tokens` is only consumed by `eval_val()` calls at validation checkpoints (line 1251) and the post-serialization roundtrip eval (line 1404). No gradient steps are taken on val_tokens at any point. ### Check 3: LEGAL score-first TTT (PR #1413 pattern, is_last_chunk guard) NOT APPLICABLE — no TTT present at all. ### Check 4: HOLD scored-region SLOT NOT APPLICABLE — no scored-region slot manipulation detected. ### Check 5: Temperature calibration `find_temp()` (lines 1043–1054) is invoked only when `TEMP_SCALING=1` (line 1409). It uses `calibration_tokens` sourced from `train_loader.stream.take(65536)` (line 1412) — train data, not val data. No val leakage. ### Architecture notes - Ternary BitNet (group_size=64), QAT with STE -...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
